src/main/java/com/zendesk/maxwell/replication/BinlogConnectorEventListener.java

```diff
@@ -44,19 +70,23 @@ public void onEvent(Event event) {
 		long eventSeenAt = 0L;
 		boolean trackMetrics = false;
 
-		if (event.getHeader().getEventType() == EventType.QUERY) {
+		EventType eventType = event.getHeader().getEventType();
+
+		if (eventType == EventType.QUERY) {
+			transactionStartTime = event.getHeader().getTimestamp();
+		} else if (eventType == EventType.XID) {
 			trackMetrics = true;
 			eventSeenAt = System.currentTimeMillis();
-			QueryEventData queryEventData = event.getData();
-			mysqlTxExecutionTimer.update(queryEventData.getExecutionTime(), TimeUnit.SECONDS);
+			long transactionCommitTime = event.getHeader().getTimestamp();
+			replicationLag = eventSeenAt - transactionCommitTime;
+			executionTime = transactionCommitTime - transactionStartTime;
```

Comment:
Do we need some sanity checking here in case it gets triggered without seeing a prior QUERY event type? (i.e. if `transactionStartTime` == 0 it's probably not legitimate). Or are we guaranteed to always see QUERY before XID?

Reply:
Good point. Will move this metric into replicator after transaction rollup.